### PR TITLE
[#45772] use direct upload tokens for direct upload

### PR DESCRIPTION
**Pull request by Kharonus**
- https://community.openproject.org/work_packages/45772
- added feature flag for the direct upload implementation in the backend
  - upload with shares are now "legacy", enable it with the feature flag legacy_upload_preparation
- use upload toasts for storage upload
- use drop zone for upload

### CODE REVIEW

File: modules/storages/app/common/storages/peripherals/storage_interaction/storage_queries.rb [hunk 1]
```diff
@@ -52,16 +52,23 @@ def download_link_query
...
       case @provider_type
       when ::Storages::Storage::PROVIDER_TYPE_NEXTCLOUD
         retry_with_refreshed_token do |token, with_refreshed_token_proc|
-          ::Storages::Peripherals::StorageInteraction::Nextcloud::UploadLinkQuery.new(
-            base_uri: @uri,
-            token:,
-            retry_proc: with_refreshed_token_proc,
-            finalize_url:
-          )
+          if OpenProject::FeatureDecisions.legacy_upload_preparation_active?
```
Comment (apfohl) [inline modules/storages/app/common/storages/peripherals/storage_interaction/storage_queries.rb:59]: Maybe we should define a point of time in the future to remove the legacy upload method. A WP or so?

File: frontend/src/app/core/file-upload/op-file-upload.service.ts [hunk 2]
```diff
@@ -76,7 +80,10 @@ export class OpenProjectFileUploadService {
   public uploadAndMapResponse(url:string, files:UploadFile[], method = 'post') {
     const { uploads, finished } = this.upload(url, files);
     const mapped = finished
-      .then((result:HalResource[]) => result.map((el:HalResource) => ({ response: el, uploadUrl: el.staticDownloadLocation.href }))) as Promise<{ response:HalResource, uploadUrl:string }[]>;
+      .then((result:HalResource[]) => result.map((el:HalResource) => ({
```
Comment (apfohl) [inline frontend/src/app/core/file-upload/op-file-upload.service.ts]: `el`? Is it `element`?

File: frontend/src/app/core/file-upload/op-file-upload.service.ts [hunk 3]
```diff
@@ -98,6 +105,7 @@ export class OpenProjectFileUploadService {
    * @param {string} url
    * @param {UploadFile} file
    * @param {string} method
+   * @param {'text'|'json'} responseType
```
Comment (apfohl) [inline frontend/src/app/core/file-upload/op-file-upload.service.ts]: I would suggest, adding the return type to the function and remove the whole comment.

File: frontend/src/app/core/state/file-links/file-links.service.ts [hunk 4]
```diff
@@ -102,19 +101,10 @@ export class FileLinksResourceService extends ResourceCollectionService<IFileLin
...
-        id: file.id,
-        name: file.name,
-        mimeType: file.mimeType,
-        size: file.size,
-        createdAt: file.createdAt,
-        lastModifiedAt: file.lastModifiedAt,
-        createdByName: file.createdByName,
-        lastModifiedByName: file.lastModifiedByName,
-      },
+      originData: { ...file },
```
Comment (apfohl) [inline frontend/src/app/core/state/file-links/file-links.service.ts:107]: 👍

File: frontend/src/app/shared/components/storages/storage/storage.component.ts [hunk 5]
```diff
@@ -127,6 +136,8 @@ export class StorageComponent extends UntilDestroyedMixin implements OnInit, OnD
       uploadFailed: (fileName:string):string => this.i18n.t('js.storages.file_links.upload_error', { fileName }),
       linkingAfterUploadFailed: (fileName:string, workPackageId:string):string =>
         this.i18n.t('js.storages.file_links.link_uploaded_file_error', { fileName, workPackageId }),
+      draggingManyFiles: (storageType:string):string => this.i18n.t('js.storages.file.dragging_many_files', { storageType }),
```
Comment (apfohl) [inline frontend/src/app/shared/components/storages/storage/storage.component.ts:139]: `draggingTooManyFiles`?

File: frontend/src/app/shared/components/storages/storage/storage.component.ts [hunk 6]
```diff
@@ -300,6 +312,58 @@ export class StorageComponent extends UntilDestroyedMixin implements OnInit, OnD
...
+    const notification = this.toastService.add({
+      data: [[file, observable]],
+      type: 'upload',
+      message: this.text.toast.uploadingLabel,
+    });
+
+    return observable
+      .pipe(
+        tap(() => {
+          setTimeout(() => this.toastService.remove(notification), 700);
```
Comment (apfohl) [inline frontend/src/app/shared/components/storages/storage/storage.component.ts:345]: Not for this PR, but wouldn't it be nice if the close timeout could be configured while creating the toast and kept track in the `ToastService`? And maybe not with `setTimeout` nut with `pipe(delay(...))`? This side effect here bothers me.

File: frontend/src/app/shared/components/toaster/upload-progress.component.ts [hunk 7]
```diff
@@ -48,85 +60,87 @@ import { UploadFile, UploadHttpEvent, UploadInProgress } from 'core-app/core/fil
...
               debugLog(`File ${this.fileName} was fully uploaded.`);
               this.value = 100;
               this.completed = true;
-              return this.onSuccess.emit();
-
+              this.uploadSuccess.emit();
+              break;
             default:
-            // Sent or unknown event
+            // unknown event
```
Comment (apfohl) [inline frontend/src/app/shared/components/toaster/upload-progress.component.ts]: Maybe add a log message as well.

**Review by apfohl — CHANGES_REQUESTED** [review]
Nothing seems to be wrong, but I still have some suggestions. Just ask me for another review, if you have changed something.

File: frontend/src/app/shared/components/storages/storage/storage.component.ts (hunk 5 above)
Comment (Kharonus) [reply frontend/src/app/shared/components/storages/storage/storage.component.ts:139]: I see no reason for making it longer. ;) `many` is quite the correct word if you want to emphasize "more the one".

File: frontend/src/app/core/file-upload/op-file-upload.service.ts (hunk 2 above)
Comment (Kharonus) [reply frontend/src/app/core/file-upload/op-file-upload.service.ts]: renamed

File: modules/storages/app/common/storages/peripherals/storage_interaction/storage_queries.rb (hunk 1 above)
Comment (Kharonus) [reply modules/storages/app/common/storages/peripherals/storage_interaction/storage_queries.rb:59]: I'd couple it to the successful usage of the real direct upload for any amount of time in production.

File: frontend/src/app/core/file-upload/op-file-upload.service.ts (hunk 3 above)
Comment (Kharonus) [reply frontend/src/app/core/file-upload/op-file-upload.service.ts]: fixed

File: frontend/src/app/shared/components/toaster/upload-progress.component.ts (hunk 7 above)
Comment (Kharonus) [reply frontend/src/app/shared/components/toaster/upload-progress.component.ts]: added a  browser console warning

File: frontend/src/app/shared/components/storages/storage/storage.component.ts (hunk 6 above)
Comment (Kharonus) [reply frontend/src/app/shared/components/storages/storage/storage.component.ts:345]: yes, sounds sensible, yet the refactoring of the toast service would exceed the current scope of this WP.

**Review by apfohl — APPROVED** [review]

**Pull request merged**